base/inference.jl

```diff
@@ -34,10 +34,18 @@ typealias VarTable Array{Any,1}
 type VarState
     typ
     undef::Bool
+    function VarState(typ::ANY, undef::Bool)
+        if isa(typ, DataType) && isdefined(typ, :instance)
```

Comment:
Yes, although that would miss the specific case I cared about (turning singleton arguments into `Const` values).